apng.cpp

```diff
@@ -0,0 +1,457 @@
+        if (memcmp(frame->ptr(i, 0), prev_frame->ptr(i, 0), 4 * frame->cols) != 0) {
+            hdr->height = i - hdr->y_offset + 1;
+            break;
+        }
+    }
+
+    // count identical cols on left
+    for (int i = 0; i < frame->cols; i++) {
+        bool differs = false;
+        for (int y = hdr->y_offset; y < hdr->y_offset + hdr->height; y++) {
+            if (memcmp(frame->ptr(y, i), prev_frame->ptr(y, i), 4) != 0) {
+                differs = true;
+                break;
+            }
+        }
+        if (differs) {
+            hdr->x_offset = i;
+            break;
+        }
+    }
+
+    // count identical cols on right
+    for (int i = frame->cols - 1; i >= 0; i--) {
+        bool differs = false;
+        for (int y = hdr->y_offset; y < hdr->y_offset + hdr->height; y++) {
+            if (memcmp(frame->ptr(y, i), prev_frame->ptr(y, i), 4) != 0) {
+                differs = true;
+                break;
+            }
+        }
+        if (differs) {
+            hdr->width = i - hdr->x_offset + 1;
+            break;
+        }
+    }
+}
+
+void apng_diff_frame(uint8_t *out, const struct apng_frame_header *hdr, const cv::Mat* frame, const cv::Mat* prev_frame) {
+    for (int y = hdr->y_offset; y < hdr->y_offset + hdr->height; y++) {
+        for (int x = hdr->x_offset; x < hdr->x_offset + hdr->width; x++) {
+            if (memcmp(frame->ptr(y, x), prev_frame->ptr(y, x), 4) == 0) {
+                // Colors match, emit transparent pixel
+                memset(out, 0, 4);
+                out += 4;
+            } else {
+                memcpy(out, frame->ptr(y, x), 3);
+                out += 3;
+                // fully opaque pixel
+                *out++ = 0xFF;
+            }
+        }
+    }
+}
+
+bool apng_encoder_encode_frame(apng_encoder e, const opencv_mat frame, int ms) {
+    auto mat = static_cast<const cv::Mat*>(frame);
+
+    if (setjmp(png_jmpbuf(e->png_ptr))) {
+        printf("jumpbuf happened\n");
+        return false;
+    }
+
+    struct apng_frame_header hdr;
+
+    // find the smallest rectangle of changed pixels
+    apng_find_diff_bounds(&hdr, mat, e->prev_frame);
+
+    // todo: handle case where new frame is partially transparent
+
+    uint8_t *buf = (uint8_t*) calloc(4, hdr.width * hdr.height);
+    // copy differing pixels into frame
+    apng_diff_frame(buf, &hdr, mat, e->prev_frame);
+
+    png_bytepp row_pointers = (png_bytepp) png_malloc(e->png_ptr, sizeof(png_bytep) * mat->rows);
+    if (!row_pointers) return false;
+    for(int i = 0; i < hdr.height; i++) {
+        row_pointers[i] = buf + i * hdr.width * 4;
+    }
+
+    png_write_frame_head(e->png_ptr, e->info_ptr, row_pointers,
+        hdr.width, hdr.height,
+        hdr.x_offset, hdr.y_offset,
+        ms, 1000, /* Delay */
+        PNG_DISPOSE_OP_NONE,
+        PNG_BLEND_OP_OVER);
+
+    png_write_image(e->png_ptr, row_pointers);
+    png_write_frame_tail(e->png_ptr, e->info_ptr);
+
+    mat->copyTo(*e->prev_frame);
```

Comment:
Is there a reason we couldn't use `mat`'s buffer instead of creating `buf` above?